File: DoubleLinkedList.java
package JAVA;

public class DoubleLinkedList{

    Node_ head = null;
    public static void main(String[] args) {
        DoubleLinkedList ll = new DoubleLinkedList();
        ll.add(1);
        ll.add(2);
        ll.add(3);
        ll.add(4);
        ll.add(5);
        ll.print_();
    }
    void add(int data){

        if(head == null){
            Node_ newNode = new Node_(null,data,null);
            head = newNode;
            return;
        }
        Node_ currentNode = head;
        while(currentNode.next != null){
            currentNode = currentNode.next;
        }
        Node_ newNode = new Node_(currentNode,data,null);
        currentNode.next = newNode;
        return;

    }

    void print_(){
        while (head != null){
            System.out.println("----------------------------------------------------------");
            System.out.println(head.previous+"--"+head.data+"--"+head.next);

            head = head.next;
        }
    }


}

class Node_{
    Node_ previous;
    int data;
    Node_ next;

    Node_(Node_ previous, int data, Node_ next){
        this.previous = previous;
        this.data = data;
        this.next = next;
    }


}

File: GO-CODE/DoubleLinkedList.go
package main

import (
	"fmt"
)

type Node struct{
	previous *Node
	data int
	next *Node
}

type doubleLinkedList struct{
	head *Node
}

func (dll *doubleLinkedList) add(data int){
	
	if(dll.head == nil){
		node := &Node{previous:nil, data:data, next:nil}
		dll.head = node
		return
	}
	
	currentNode := dll.head
	for(currentNode.next != nil){
		currentNode = currentNode.next
	}
	node := &Node{previous:currentNode, data:data, next:nil}
	currentNode.next = node
	return

	
}

func (dll *doubleLinkedList) Display (){
	headPointer := dll.head
	for(dll.head != nil){
		fmt.Println(dll.head)
		dll.head = dll.head.next
	}
	dll.head = headPointer
}


func (dll *doubleLinkedList) addAtIndex(data int, idx int){
	saveheadPointer := dll.head
	currentNode := dll.head
	var index int = 0
	
	for(currentNode != nil){
		
		if(index == idx){
			
			node := &Node{previous:currentNode.previous, data:data, next:currentNode.previous.next}
			currentNode.previous.next = node
			currentNode.previous = node
			currentNode = node
			break
		}
		index += 1
		currentNode = currentNode.next
		
	}
	
	dll.head = saveheadPointer
}

func main(){
	dll := &doubleLinkedList{}
	dll.add(1)
	dll.add(2)
	dll.add(3)
	dll.add(4)
	dll.Display()
	dll.addAtIndex(5,2)
	dll.Display()
}